state/accumulator_test.go

```diff
@@ -618,6 +619,140 @@ func TestAccumulatorConcurrency(t *testing.T) {
 	}
 }
 
+// Sanity-check the accumulator's logic for inserting missing_previous markers.
+func TestAccumulatorMissingPreviousMarkers(t *testing.T) {
+	db, close := connectToDB(t)
+	defer close()
+	accumulator := NewAccumulator(db)
+
+	t.Log("Initialise a room.")
+	roomID := fmt.Sprintf("!%s:localhost", t.Name())
+	initialEvents := []json.RawMessage{
+		[]byte(`{"event_id":"$state-A", "type":"m.room.create", "state_key":"", "content":{"creator":"@me:localhost"}}`),
+		[]byte(`{"event_id":"$state-B", "type":"m.room.member", "state_key":"@me:localhost", "content":{"membership":"join"}}`),
+		[]byte(`{"event_id":"$state-C", "type":"m.room.join_rules", "state_key":"", "content":{"join_rule":"public"}}`),
+	}
+	_, err := accumulator.Initialise(roomID, initialEvents)
+	if err != nil {
+		t.Fatalf("failed to Initialise accumulator: %s", err)
+	}
+
+	steps := []struct {
+		Desc            string
+		Events          []json.RawMessage
+		Limited         bool
+		NumNew          int
+		CheckDesc       string
+		MissingPrevious map[string]bool
```

Comment:
Please be clearer on the inputs/outputs, otherwise it's hard to see what is case data and what is assertion output.